CONTRIBUTING: update release checklist (#2238)
XRPLF/xrpl.js: CONTRIBUTING.md

npm uninstall abbrev -w xrpl
```

## Release process + checklist

## Pre-release
- [ ] Your changes should be on a branch.
- [ ] Your changes should have unit tests.
- [ ] Lint the code with `npm lint`
- [ ] Build your code with `npm build`
- [ ] Run the unit tests with `npm test`
- [ ] Get a full code review.
- [ ] Merge your branch into `main` and push to github.
- [ ] Ensure that all tests passed on the last CI that ran on `main`.

## Release

> - [ ] Checkout `main` (or your beta branch) and `git pull`.
> - [ ] Create a new branch (`git checkout -b <BRANCH_NAME>`) to capture updates that take place during this process.
> - [ ] Update `HISTORY.md` to reflect release changes.
>   - [ ] Update the version number and release date, and ensure it lists the changes since the previous release.
> - [ ] Run `npm run docgen` if the docs were modified in this release to update them (skip this step for a beta).
> - [ ] Run `npm run build` to triple check the build still works
> - [ ] Run `npx lerna version --no-git-tag-version` - This creates a draft PR and bumps the versions of the packages.
>   - [ ] For each changed package, pick what the new version should be. Lerna will bump the versions, commit version bumps to `main`, and create a new git tag for each published package.
>   - [ ] If publishing a beta, make sure that the versions are all of the form `a.b.c-beta.d`, where `a`, `b`, and `c` are identical to the last normal release except for one, which has been incremented by 1.
> - [ ] Run `npm i` to update the package-lock with the updated versions.
> - [ ] Create a new PR from this branch into `main` and merge it (you can directly merge into the beta branch for a beta).
> - [ ] Checkout `main` and `git pull` (you can skip this step for a beta since you already have the latest version of the beta branch).
> - [ ] Actually publish the packages with one of the following:
>   - [ ] Stable release: Run `npx lerna publish from-package --yes`
>   - [ ] Beta release: Run `npx lerna publish from-package --dist-tag beta --yes`
>         Notice this allows developers to install the package with `npm add xrpl@beta`
>   - [ ] If requested, enter your [npmjs.com](https://npmjs.com) OTP (one-time password) to complete publication.
> - [ ] If not a beta release: Create a new branch (`git checkout -b <BRANCH_NAME>`) to capture the updated packages from the release. Merge those changes into `main`.

Reviewer: Re-adding numbers using auto numbering since checkboxes aren't dynamic, and with all the alt-tabbing needed it's easy to lose track of where you are unless you have the numbers. 
```suggestion
1. Checkout `main` (or your beta branch) and `git pull`.
1. Create a new branch (`git checkout -b <BRANCH_NAME>`) to capture updates that take place during this process.
1. Update `HISTORY.md` to reflect release changes.

- [ ] Update the version number and release date, and ensure it lists the changes since the previous release.

1. Run `npm run docgen` if the docs were modified in this release to update them (skip this step for a beta).
1. Run `npm run build` to triple check the build still works
1. Run `npx lerna version --no-git-tag-version` - This creates a draft PR and bumps the versions of the packages.

- [ ] For each changed package, pick what the new version should be. Lerna will bump the versions, commit version bumps to `main`, and create a new git tag for each published package.
- [ ] If publishing a beta, make sure that the versions are all of the form `a.b.c-beta.d`, where `a`, `b`, and `c` are identical to the last normal release except for one, which has been incremented by 1.

1. Run `npm i` to update the package-lock with the updated versions.
1. Create a new PR from this branch into `main` and merge it (you can directly merge into the beta branch for a beta).
1. Checkout `main` and `git pull` (you can skip this step for a beta since you already have the latest version of the beta branch).
1. Actually publish the packages with one of the following:

- [ ] Stable release: Run `npx lerna publish from-package --yes`
- [ ] Beta release: Run `npx lerna publish from-package --dist-tag beta --yes`
      Notice this allows developers to install the package with `npm add xrpl@beta`
- [ ] If requested, enter your [npmjs.com](https://npmjs.com) OTP (one-time password) to complete publication.

1. If not a beta release: Create a new branch (`git checkout -b <BRANCH_NAME>`) to capture the updated packages from the release. Merge those changes into `main`.
```

npm uninstall abbrev -w xrpl
```

## Release process + checklist

> ## Pre-release

Reviewer 2: IMO this should be labeled as the process for a PR instead of for a release

> ## Pre-release

Author: ```suggestion
## PR process
```

1. If not a beta release: Create a new branch (`git checkout -b <BRANCH_NAME>`) to capture the updated packages from the release. Merge those changes into `main`.

   NOW YOU HAVE PUBLISHED! But you're not done; we have to notify people!

1. Pull the most recent changes to `main` locally.
1. Run `git tag <tagname> -m <tagname>`, where `<tagname>` is the new package and version (e.g. `xrpl@2.1.1`), for each version released.
1. Run `git push --follow-tags`, to push the tags to Github.
1. On GitHub, click the "Releases" link on the right-hand side of the page.

1. Repeat for each release:

   1. Click "Draft a new release"
   1. Click "Choose a tag", and choose a tag that you just created.
   1. Edit the name of the release to match the tag (IE \<package\>@\<version\>) and edit the description as you see fit.

1. Lastly, send an email to [xrpl-announce](https://groups.google.com/g/xrpl-announce).

# ripple-lib 1.x releases

- [ ] Publish the release to npm.

  - [ ] If you are publishing a 1.x release to the `xrpl` package, use:

        npm publish --tag ripple-lib

    This prevents the release from taking the `latest` tag.

> For ripple-lib, push to both `ripple-lib` and `xrpl` (npm package names). For xrpl, use `--tag ripple-lib` for ripple-lib releases; [here's why](https://blog.greenkeeper.io/one-simple-trick-for-javascript-package-maintainers-to-avoid-breaking-their-user-s-software-and-to-6edf06dc5617).

Author: for the npm package? It looks like maintainers are:

ahbritto
justmoon
geertweening
wilsonianb
shekenahglory
intelliot
hansbergren

Unfortunately, it doesn't look like I have permission to add maintainers. Though I could be wrong about this, as the npmjs interface isn't clear.

> For ripple-lib, push to both `ripple-lib` and `xrpl` (npm package names). For xrpl, use `--tag ripple-lib` for ripple-lib releases; [here's why](https://blog.greenkeeper.io/one-simple-trick-for-javascript-package-maintainers-to-avoid-breaking-their-user-s-software-and-to-6edf06dc5617).

Reviewer 2: Yes, I remember that being an issue the last time we dealt with this.

> For ripple-lib, push to both `ripple-lib` and `xrpl` (npm package names). For xrpl, use `--tag ripple-lib` for ripple-lib releases; [here's why](https://blog.greenkeeper.io/one-simple-trick-for-javascript-package-maintainers-to-avoid-breaking-their-user-s-software-and-to-6edf06dc5617).

Author: Alright, I'll add my name in the doc for now

> For ripple-lib, push to both `ripple-lib` and `xrpl` (npm package names). For xrpl, use `--tag ripple-lib` for ripple-lib releases; [here's why](https://blog.greenkeeper.io/one-simple-trick-for-javascript-package-maintainers-to-avoid-breaking-their-user-s-software-and-to-6edf06dc5617).

Author: ```suggestion
For ripple-lib:

 - Have one of the ripple-lib package maintainers push to `ripple-lib` (npm package name). You can contact [@intelliot](https://github.com/intelliot) to request the npm publish.
- For ripple-lib releases, cross-publish the package to `xrpl` with `--tag ripple-lib`
  - [Here's why](https://blog.greenkeeper.io/one-simple-trick-for-javascript-package-maintainers-to-avoid-breaking-their-user-s-software-and-to-6edf06dc5617).
```
